{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "04a6d658",
   "metadata": {},
   "source": [
    "## Spatial Transcriptomics simulation with spatial dependencies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e50b8db",
   "metadata": {},
   "source": [
    "- **Fix library-size sampling noise** <br>\n",
    "We sample single cell indices with similar library size to the same synthetic spots. <br><br>\n",
    "\n",
    "- **Add spatial dependencies with 2D Gaussian Process simulations**<br>\n",
    "There is smooth transition of both cell-type specific and overall spatial density patterns (e.g. similar density for adjacent spots)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "942b334a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import math\n",
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import pymc3 as pm\n",
    "\n",
    "import scanpy as sc\n",
    "import torch\n",
    "import gpytorch\n",
    "\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ba728fb",
   "metadata": {},
   "source": [
    "### Cell2Location simulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c0a354a",
   "metadata": {},
   "source": [
    "**Step 1: simulate per cell abundance for each spatial pattern**\n",
    "- $r$: Spatial patterns: (1). ubiquitous ($U = \\{u_{1},\n",
    "\\dots,u_{n_1}\\}$); (2). regional ($R = \\{r_{1},\n",
    "\\dots,r_{n_2}\\}$) (consisting of sub-patterns of number $n_1$, $n_2$)<br>\n",
    "- $f$: cell type (in our case $f=9$<br>\n",
    "- $a$: abundance (density): (1). high density (H); (2). low density (L)<br>\n",
    "- $x_{r,f}$: binary assignment of cell types to spatial patterns\n",
    "    - e.g. $x_{r=U, f=k} = 0$:  cell type $k$ exist in ubiquitous spatial pattern\n",
    "\n",
    "- $d_f$: per cell type average abundance: \n",
    "    - $d_f^{r=U, a=L} \\sim \\text{Gamma}(5, 5)$\n",
    "    - $d_f^{r=U, a=H} \\sim \\text{Gamma}(14, 5)$\n",
    "    - $d_f^{r=R, a=L} \\sim \\text{Gamma}(5, 5)$\n",
    "    - $d_f^{r=R, a=H} \\sim \\text{Gamma}(14, 5)$\n",
    "\n",
    "- $s$: spatial coordinates (e.g. $s = (50, 50)$)<br>\n",
    "- $z_{s,r}$: spatial abundance for spatial patterns $r$\n",
    "    - $z_{s, r=U} \\sim  \\text{GP}(\\mu=0, \\eta=0.5, \\text{bw}=bw_{r})$\n",
    "    - $z_{s, r=R} \\sim \\text{GP}(\\mu=0, \\eta=1.5, \\text{bw}=bw_{r})$<br>\n",
    "    (high $\\eta$ $\\rightarrow$ high GP variance, $bw_{r} \\sim \\text{Gamma}(9.6, 1.2)$<br>\n",
    "    $z_{s,r} = \\text{softmax}(z_{s,r}) = \\dfrac{ e^{z_{s,r}} }{ \\sum e^{z_{s,r}}  }$  - ensure positive scale<br>\n",
    "    $z_{s,r} = \\dfrac{z_{s,r}}{\\max_s z_{s,r}}$\n",
    "\n",
    "- $w_{s,f}$: Per cell-type abundance for each location $s$<br>\n",
    "    - $w_{s,f} = \\sum_{r} (z_{s,r} x_{r,f}) \\cdot q_{s,f}$ <br>\n",
    "    $q_{s,f} \\sim \\text{LogNormal}(\\mu=0, \\sigma=0.35)$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66bfc98e",
   "metadata": {},
   "source": [
    "Note:\n",
    "exponentiated quadratic kernel: $k(x_1, x_2) = \\sigma^2 \\exp{- \\dfrac{\\Vert x_1 - x_2 \\Vert^2}{2 l^2}}$\n",
    "- $\\sigma^2$: overall variance (amplitude)\n",
    "- $l$: lengthscale"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be3ad404",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code from cell2location simulations\n",
    "# Reference: https://github.com/vitkl/cell2location_paper\n",
    "\n",
    "def kernel(X1, X2, l=1.0, eta=1.0):\n",
    "    \"\"\"\n",
    "    Isotropic squared exponential kernel. Computes \n",
    "    a covariance matrix from points in X1 and X2.\n",
    "    \n",
    "    Args:\n",
    "        X1: Array of m points (m x d).\n",
    "        X2: Array of n points (n x d).\n",
    "\n",
    "    Returns:\n",
    "        Covariance matrix (m x n).\n",
    "    \"\"\"\n",
    "    sqdist = np.sum(X1**2, 1).reshape(-1, 1) + np.sum(X2**2, 1) - 2 * np.dot(X1, X2.T)\n",
    "    return eta**2 * np.exp(-0.5 / l**2 * sqdist)\n",
    "    \n",
    "def generate_grid(n=[50, 50]):\n",
    "    n1, n2 = n \n",
    "    x1 = np.linspace(0, 100, n1)[:,None] #saptial dimensions \n",
    "    x2 = np.linspace(0, 100, n2)[:,None] #saptial dimensions \n",
    "\n",
    "    # make cartesian grid out of each dimension x1 and x2\n",
    "    return pm.math.cartesian(x1[:,None], x2[:,None]), x1, x2\n",
    "\n",
    "def random_GP(X, \n",
    "              x1=1, x2=1, #coordinates\n",
    "              n_variables = 5, # zones\n",
    "              eta_true = 5, #variance, defines overlapping\n",
    "              l1_true=[8, 10, 15], #bw parameter\n",
    "              l2_true=[8, 10, 15]\n",
    "             ):\n",
    "\n",
    "    #cov1, cov2 = kernel(x1, x1, l=l1_true), kernel(x2, x2, l=l2_true)\n",
    "    K = [np.kron(kernel(x1, x1, l=l1_true[i], eta=eta_true), \n",
    "                 kernel(x2, x2, l=l2_true[i], eta=eta_true)) \n",
    "         for i in range(n_variables)]\n",
    "    \n",
    "    #samples from GP\n",
    "    gaus_true = np.stack([\n",
    "        np.random.multivariate_normal(np.zeros(X.shape[0]), 2*K[i]) \n",
    "        for i in range(n_variables)\n",
    "    ]).T \n",
    "    \n",
    "    N_true = (np.exp(gaus_true).T / np.exp(gaus_true).sum(axis=1)).T #softmax transform \n",
    "    return N_true\n",
    "\n",
    "\n",
    "# TODO:\n",
    "# Try increasing GP eta (variance) value --> more inhomogenous cell-type density \n",
    "def sample_GP(locations, \n",
    "              ct_labels,\n",
    "              n_spec,\n",
    "              n_ubiq,\n",
    "              bw_spec,\n",
    "              bw_ubiq,\n",
    "              x1, \n",
    "              x2,\n",
    "              eta_spec=2,\n",
    "              eta_ubiq=1\n",
    "             ):\n",
    "    \"\"\"\n",
    "    Sample abundances with GP\n",
    "    \n",
    "        Higher `eta` values represent higher transition rates\n",
    "        from high-density <--> low -density region\n",
    "        \n",
    "    \"\"\"\n",
    "    sparse_abundances = random_GP(\n",
    "        X=locations, \n",
    "        x1=x1, \n",
    "        x2=x2, \n",
    "        n_variables=n_spec, \n",
    "        eta_true=eta_spec,\n",
    "        l1_true=bw_spec,\n",
    "        l2_true=bw_spec\n",
    "    )\n",
    "    \n",
    "    sparse_abundances = sparse_abundances / sparse_abundances.max(0)\n",
    "    sparse_abundances[sparse_abundances < 0.1] = 0\n",
    "\n",
    "    uniform_abundances = random_GP(\n",
    "        X=locations, \n",
    "        x1=x1, \n",
    "        x2=x2, \n",
    "        n_variables=n_ubiq, \n",
    "        eta_true=eta_ubiq,\n",
    "        l1_true=bw_ubiq,\n",
    "        l2_true=bw_ubiq\n",
    "    )\n",
    "    \n",
    "    uniform_abundances = uniform_abundances / uniform_abundances.max(0)\n",
    "    uniform_abundances[uniform_abundances < 0.1] = 0\n",
    "\n",
    "    abundances = np.concatenate([sparse_abundances, uniform_abundances], axis=1)\n",
    "    \n",
    "    return pd.DataFrame(\n",
    "        abundances, \n",
    "        index=[f'location_{i}' for i in range(abundances.shape[0])],\n",
    "        columns=ct_labels\n",
    "    )\n",
    "\n",
    "\n",
    "def plot_spatial(values, n=[50,50], nrows=5, names=['cell type'],\n",
    "                 vmin=0, vmax=1):\n",
    "    \n",
    "    n_cell_types = values.shape[1]\n",
    "    n1, n2 = n \n",
    "    ncols = np.ceil((n_cell_types+1) / nrows).astype(np.uint16)\n",
    "    for ct in range(n_cell_types):\n",
    "        try:\n",
    "            plt.subplot(nrows, ncols, ct+1)\n",
    "            plt.imshow(values[:,ct].reshape(n1,n2).T, \n",
    "                       cmap=plt.cm.get_cmap('magma'),\n",
    "                       vmin=vmin, vmax=vmax\n",
    "                      )\n",
    "            plt.colorbar()\n",
    "            if len(names) > 1:\n",
    "                plt.title(names[ct])\n",
    "            else:\n",
    "                plt.title(f'{names[0]} {ct+1}')\n",
    "                \n",
    "        except ValueError:\n",
    "            continue\n",
    "\n",
    "    try:\n",
    "        plt.subplot(nrows, ncols, n_cell_types+1) \n",
    "        plt.imshow(values.sum(axis=1).reshape(n1,n2).T, \n",
    "                   cmap=plt.cm.get_cmap('Greys'))\n",
    "        plt.colorbar()\n",
    "        plt.title('total')\n",
    "    \n",
    "    except ValueError as e:\n",
    "        print(e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9c0cb1d",
   "metadata": {},
   "source": [
    "**Modification:** <br>\n",
    "Direct assignment of ubiquitous & specific cell types:\n",
    "- Ubiquitous: B-cell, T-cell, Endothelial, Plasmablast\n",
    "- Specific: Normal_Epithelial, Cancer_Epithelial, CAFs, PVL, Myeloid\n",
    "\n",
    "(Note: unify cell-type names: Cancer_Epithelial = TNBC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20dd99cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ct_name(ct_map, idxs):\n",
    "    try:\n",
    "        return [ct_map[i] for i in idxs]\n",
    "    except KeyError(e):\n",
    "        print('Index {} not included in cell type map'.format(i))\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc20b072",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define cell types & cell-type assignment to spatial patterns\n",
    "ct_names = [\n",
    "    'B-cells',\n",
    "    'CAFs',\n",
    "    'Endothelial',\n",
    "    'Normal_Epithelial',\n",
    "    'Myeloid',\n",
    "    'PVL',\n",
    "    'Plasmablasts',\n",
    "    'T-cells',\n",
    "    'Cancer_Epithelial'\n",
    "]\n",
    "\n",
    "ubiq_ct_names = [\n",
    "    'B-cells',\n",
    "    'Endothelial',\n",
    "    'T-cells',\n",
    "    'Plasmablasts'\n",
    "]\n",
    "\n",
    "spec_ct_names = [\n",
    "    'CAFs',\n",
    "    'Normal_Epithelial',\n",
    "    'Myeloid',\n",
    "    'PVL',\n",
    "    'Cancer_Epithelial'\n",
    "]\n",
    "\n",
    "ct_maps = {i: ct for (i, ct) in enumerate(ct_names)}\n",
    "\n",
    "n_cts = len(ct_names) # number of cell types\n",
    "n_spec_cts = len(spec_ct_names)\n",
    "n_ubiq_cts = len(ubiq_ct_names)\n",
    "\n",
    "spec_cell_types = [\n",
    "    idx\n",
    "    for idx, name in enumerate(ct_names)\n",
    "    if name in set(spec_ct_names)\n",
    "]\n",
    "\n",
    "ubiq_cell_types = [\n",
    "    idx\n",
    "    for idx, name in enumerate(ct_names)\n",
    "    if name in set(ubiq_ct_names)\n",
    "]\n",
    "\n",
    "# numerical representation of cell types (NOTE: order super important!)\n",
    "cell_types = np.concatenate([spec_cell_types, ubiq_cell_types])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c95ec4c",
   "metadata": {},
   "source": [
    "**TODO**\n",
    "\n",
    "Simulate whole canvas and mask out tissue regions with true spatial regions (simualate hotspot of specific cell types (using localized patterns).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d11cd8e",
   "metadata": {},
   "source": [
    "Load image & spot coordinates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5f6bc09",
   "metadata": {},
   "outputs": [],
   "source": [
    "from skimage import io\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87b23964",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Please re-specify the spatial data path\n",
    "# Data: CID44971 \n",
    "spatial_path = '../../data/wu_data/CID44971/st/'\n",
    "sample_id = 'CID44971'\n",
    "\n",
    "\n",
    "# Load real ST data (CID44971)\n",
    "adata = sc.read_h5ad(\n",
    "    os.path.join(spatial_path, sample_id+'.h5ad')\n",
    ")\n",
    "\n",
    "adata.var_names_make_unique()\n",
    "\n",
    "# Load image & locations\n",
    "\"\"\"\n",
    "with open('../../data/1a/spatial/scalefactors_json.json', 'r') as f:\n",
    "    scale_info = json.load(f)\n",
    "    scale = scale_info['tissue_hires_scalef']\n",
    "    \n",
    "locs = np.round(adata.obsm['spatial'] * scale).astype(int)\n",
    "size = hist_img.shape\n",
    "\"\"\"\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c38194a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _random_assign(n):\n",
    "    \"\"\"Random one-to-one assignemnt by permuting an NxN identity matrix\"\"\"\n",
    "    mat = np.random.permutation(np.diag(np.ones(n)))\n",
    "    return mat\n",
    "\n",
    "def assign_ct_pattern(ubiq_cts, \n",
    "                      spec_cts, \n",
    "                      n_spec_patterns,\n",
    "                      n_ubiq_patterns, \n",
    "                     ):\n",
    "    \"\"\"\n",
    "    Assign cell-type categories to spatial pattern categories\n",
    "    \"\"\"\n",
    "    n_ubiqs = len(ubiq_cts) # number of ubiquitous cell types\n",
    "    n_specs = len(spec_cts) # number of specific cell types\n",
    "    \n",
    "    # Modification:\n",
    "    # For both ubiquitous & spatial pattern, only allow one-to-one mapping\n",
    "    spec_assignment = _random_assign(n_spec_patterns)\n",
    "    ubiq_assignment = _random_assign(n_ubiq_patterns)\n",
    "    \n",
    "    \"\"\"\n",
    "    ubiq_assignment = np.random.binomial(\n",
    "        n=1,\n",
    "        p=0.4,\n",
    "        size=(len(ubiq_cts), n_ubiq_patterns)\n",
    "    )\n",
    "    \"\"\"\n",
    "    \n",
    "    mat_spec = np.concatenate(\n",
    "        (spec_assignment, \n",
    "         np.zeros((len(spec_cts), n_ubiq_patterns))\n",
    "        ),\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "    mat_ubiq = np.concatenate(\n",
    "        (np.zeros((len(ubiq_cts), n_spec_patterns)),\n",
    "         ubiq_assignment\n",
    "        ),\n",
    "        axis=1\n",
    "    )\n",
    "    \n",
    "    idx_names = np.concatenate((spec_cts, ubiq_cts))\n",
    "    col_names = ['Spec_'+str(i+1) for i in range(n_spec_patterns)] + \\\n",
    "                ['Ubiq_'+str(i+1) for i in range(n_ubiq_patterns)]\n",
    "    \n",
    "    \n",
    "    patterns_df = pd.DataFrame(\n",
    "        np.concatenate(\n",
    "            (mat_spec, mat_ubiq),\n",
    "            axis=0\n",
    "        ).astype(np.uint8),\n",
    "        index=idx_names,\n",
    "        columns=col_names\n",
    "    )\n",
    "    \n",
    "    patterns_df.index.name='cell_types'\n",
    "    \n",
    "    return patterns_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b975eda",
   "metadata": {},
   "source": [
    "Increase GP bandwidth:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c58ef639",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate matrix of which cell types are in which zones\n",
    "# Specify number of distinct spatial patterns under each category\n",
    "\"\"\"\n",
    "    Assignment of cell-type to specific spatial patterns \n",
    "    (either `Spec` - specific or `Ubiq` - Ubiquitous)\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "np.random.seed(1000)\n",
    "n_experiments = 1\n",
    "n_locations = [50, 50]\n",
    "\n",
    "ct_patterns_df = assign_ct_pattern(\n",
    "    ubiq_cell_types,\n",
    "    spec_cell_types,\n",
    "    n_spec_patterns=n_spec_cts,\n",
    "    n_ubiq_patterns=n_ubiq_cts\n",
    ")\n",
    "\n",
    "# Sample bandwidth parameter\n",
    "mean_var_ratio = 1.5\n",
    "mean_spec = 12\n",
    "mean_ubiq = 8\n",
    "\n",
    "bandwidth_spec = np.random.gamma(\n",
    "    mean_spec * mean_var_ratio, 1 / mean_var_ratio,\n",
    "    size=n_spec_cts\n",
    ")\n",
    "\n",
    "bandwidth_ubiq = np.random.gamma(\n",
    "    mean_ubiq * mean_var_ratio, 1 / mean_var_ratio,\n",
    "    size=n_ubiq_cts\n",
    ")\n",
    "\n",
    "locations_1, x1, x2 = generate_grid(n=n_locations)\n",
    "locations = np.concatenate([locations_1 for _ in range(n_experiments)], axis=0)\n",
    "\n",
    "\n",
    "# Sample per-cell-type, per-spot abundance\n",
    "abundances_df = pd.DataFrame()\n",
    "for e in range(n_experiments):\n",
    "    \n",
    "    abundances_df_1 = sample_GP(\n",
    "        locations=locations_1, \n",
    "        ct_labels=cell_types,\n",
    "        n_spec=n_spec_cts,\n",
    "        n_ubiq=n_ubiq_cts,\n",
    "        bw_spec=bandwidth_spec,\n",
    "        bw_ubiq=bandwidth_ubiq,\n",
    "        x1=x1, x2=x2\n",
    "    )\n",
    "    \n",
    "    abundances_df_1.index = [f'exper{e}_{l}' for l in abundances_df_1.index]\n",
    "    abundances_df = pd.concat((abundances_df, abundances_df_1), axis=0)\n",
    "    \n",
    "    plt.figure(figsize=(2*5+5,3*5+5))\n",
    "    plot_spatial(abundances_df_1.values, n=n_locations, nrows=5, names=np.arange(n_cts+1))\n",
    "    plt.tight_layout()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5e0d6cb",
   "metadata": {},
   "source": [
    "Generate per-spot spatial pattern:<br>\n",
    "For each cell type:\n",
    "- matrix dot spatial pattern `abundance_df` with binary assignment of spatial patterns `ct_patterns_df`\n",
    "<br><br>\n",
    "    Variable notation:\n",
    "\n",
    "    - `abundance_df` ($\\mathbb{R}^{s \\times f}$): spatial abundance for each pattern\n",
    "    - `ct_patterns_df`: binary assignment of which cell type belongs to which spatial pattern\n",
    "    - `cell_abundances_df` ($\\mathbb{R}^{s \\times f}$): spatial abundance for each cell type & spot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d49bad40",
   "metadata": {},
   "source": [
    "Latest modification: <br>\n",
    "For each cell type, we simulate independent $q_{s,c,f}$, the multiplier to generate cell abundance from spatial pattern assigment (0, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e1fd0a1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "np.random.seed(1000)\n",
    "\n",
    "cell_abundances = np.dot(abundances_df, ct_patterns_df.T)\n",
    "\n",
    "\n",
    "q_sf = np.tile(\n",
    "    np.random.lognormal(0.5, 0.5, size=n_cts),\n",
    "    reps=(n_locations[0] * n_locations[1], 1),\n",
    ")\n",
    "\n",
    "cell_abundances = cell_abundances * q_sf\n",
    "cell_abundances_df = pd.DataFrame(\n",
    "    cell_abundances, \n",
    "    index=abundances_df.index,\n",
    "    columns=ct_patterns_df.index\n",
    ")\n",
    "\n",
    "cell_abundances_df.columns = spec_ct_names + ubiq_ct_names\n",
    "\n",
    "plt.figure(figsize=(3*5+5, 3*10+5))\n",
    "plot_spatial(cell_abundances_df.values, \n",
    "             n=n_locations, nrows=5, \n",
    "             names=cell_abundances_df.columns, vmax=None)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2054c1fa",
   "metadata": {},
   "source": [
    "Discretize spot-wise cell count from cell_abundances by taking `ceiling`<br>\n",
    "Generate ground-truth proportions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fca4b35",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_count_df = np.ceil(cell_abundances_df).astype(np.uint16)\n",
    "proportion_df = cell_count_df / np.expand_dims(cell_count_df.sum(1), 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61d997ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_count_df.sum(1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bdc0886",
   "metadata": {},
   "outputs": [],
   "source": [
    "proportion_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00b64275",
   "metadata": {},
   "source": [
    "**Step 2** \n",
    "\n",
    "Sample cell indices from scRNA-seq for each spot, construct synthetic `(S x G)` counts by sampling cells with similar library size."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac61de0a",
   "metadata": {},
   "source": [
    "(1). Load scRNA-seq matrix counts & metadata (cell-type annotation)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7409c615",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc_path = '../../sc_transcriptome_ana/data/wu_data/CID44971/scrna/'\n",
    "\n",
    "# scRNA-seq count matrix\n",
    "df_sc = pd.read_csv(\n",
    "    os.path.join(sc_path, 'bc_sc.csv'),\n",
    "    index_col=[0],\n",
    "    header=[0]\n",
    ")\n",
    "\n",
    "# scRNA-seq metadata\n",
    "df_meta = pd.read_csv(\n",
    "    os.path.join(sc_path, 'metadata.csv'), index_col=[0]\n",
    ")\n",
    "\n",
    "annots = df_meta['celltype_major']\n",
    "df_sc.index.name = 'Barcode'\n",
    "df_sc['cell_type'] = annots.to_list()\n",
    "df_sc.set_index('cell_type', inplace=True, append=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02091ddd",
   "metadata": {},
   "outputs": [],
   "source": [
    "annots.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b93fd1b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac147f16",
   "metadata": {},
   "source": [
    "(2). Fit raw library size with $\\text{LogNormal}$ distribution:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f20084bc",
   "metadata": {},
   "source": [
    "**Fix inhomogeneous library size for spots captured by same synthetic spots**\n",
    "\n",
    "- (i). Fit the library size of provided ST data with `LogNormal` r.v.\n",
    "- (ii). Sample each synthetic spot $s$ with a library size $l_s$ draw from `LogNormal` r.v. retrieved from step (1)\n",
    "- (iii). For each single-cell assignment to the given synthetic spot $s$, sample the library size with poisson distribution: $l_{s_{i}} \\sim \\text{Poisson}(l_s)$\n",
    "    - For now, we don't sample per-cell library size for each spot with Poisson r.v., fix the expected library size for all scRNA-seq indices in spot $s$ as $l_s$ \n",
    "- (iv). For each single-cell assignment, and it's library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b0c4423",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import lognorm, poisson, gamma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f78eea1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fit library size r.v. with scRNA-seq data\n",
    "# plot the original histogram with the simulated one from fitted parameters\n",
    "\n",
    "libsize = df_sc.values.sum(1)\n",
    "s, loc, scale = lognorm.fit(libsize)\n",
    "\n",
    "rvs = lognorm.rvs(s, loc=loc, scale=scale, size=1000)\n",
    "plt.figure(figsize=(10, 6))\n",
    "\n",
    "plt.hist(libsize, bins=100, density=True, label='Raw libsize',\n",
    "         color='blue', edgecolor='white', alpha=0.5)\n",
    "plt.hist(rvs, bins=100, density=True, label='Fitted libsize',\n",
    "         color='green', edgecolor='white', alpha=0.5)\n",
    "\n",
    "x = np.linspace(libsize.min(), libsize.max(), len(libsize)+1)\n",
    "pdf = lognorm.pdf(x, s, loc, scale)\n",
    "plt.plot(x, pdf, '--', color='r')\n",
    "plt.legend()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d92d070f",
   "metadata": {},
   "source": [
    "(3). Generate spot x gene matrix by synthesizing `pseudo-spots` with cell-type specific density & library-size "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74517d28",
   "metadata": {},
   "source": [
    "Some visualization (archived):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e686662",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize cell count per cell-type \n",
    "plt.figure(figsize=(3*5+5, 3*10+5))\n",
    "\n",
    "plot_spatial(cell_count_df.values, \n",
    "             n=n_locations, nrows=4, \n",
    "             names=cell_abundances_df.columns, vmax=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d09fa1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize per cell-type capture rate\n",
    "eps = 1e-10\n",
    "\n",
    "cell_capture_rate_df = cell_abundances_df / (cell_count_df + eps)\n",
    "\n",
    "plt.figure(figsize=(3*5+5, 3*10+5))\n",
    "plot_spatial(cell_capture_rate_df.values,\n",
    "             n=n_locations, nrows=4,\n",
    "             names=cell_abundances_df.columns, vmax=None)\n",
    "\n",
    "plt.suptitle('Capture rate per cell type', y=0.9, fontsize=20)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c9bb78d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize avg. cell count per spot\n",
    "fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(11, 5))\n",
    "\n",
    "sns.heatmap(cell_count_df[cell_count_df>0].mean(1).values.reshape(50, -1), ax=ax1)\n",
    "ax1.set_title('Avg. non-empty cell count per spot')\n",
    "ax1.axis('off')\n",
    "\n",
    "sns.heatmap(cell_count_df[cell_count_df>0].var(1).values.reshape(50, -1), ax=ax2)\n",
    "ax2.set_title('Variance of non-empty cell count per spot')\n",
    "ax2.axis('off')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "615ccb03",
   "metadata": {},
   "source": [
    "Simulate library size for each spot:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff6fd1a4",
   "metadata": {},
   "source": [
    "First, Check overall capture ratio from scRNA-seq to visium (CID44971 scRNA - CID44971 ST) <br>\n",
    "\n",
    "We scale the true capture ratio from scRNA-seq to ST as follows:<br>\n",
    "Median library size ratio ($r = \\dfrac{\\text{ST library size}}{\\text{scRNA-seq library size}}$) divided by average `# cells` captured within each synthetic spot:<br>\n",
    "\n",
    "$$r_{scaled} = \\dfrac{r}{\\text{avg. # cells captured}}$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9565bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "libsize = df_sc.values.sum(1)\n",
    "libsize_st = adata.X.A.sum(1)\n",
    "\n",
    "cap_ratio = np.median(np.log1p(libsize_st)) / np.median(np.log1p(libsize))\n",
    "cap_ratio_per_spot = cap_ratio / cell_count_df.sum(1).mean()\n",
    "cap_ratio_per_spot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aaf6c7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_nonzero_cts(df):\n",
    "    \"\"\"\n",
    "    Find expressed (non-zero) cell types per spot\n",
    "    \"\"\"\n",
    "    cols = df.columns\n",
    "    return (df > 0).apply(\n",
    "        lambda x: list(cols[x.values]), axis=1\n",
    "    )\n",
    "\n",
    "\n",
    "def find_sc_index(df, cell_type, libsize_raw, l, n_nbr=3):\n",
    "    \"\"\"\n",
    "    Find the scRNA-seq cell with close library size to given expected library size of \n",
    "    the synthetic spot (s). \n",
    "    \n",
    "    Either randomly sample the index out of n closest libsize or \n",
    "    deterministically choose the cell index with the closest libsize\n",
    "    \n",
    "    Returns\n",
    "    -------\n",
    "    expr : list\n",
    "        Gene expression of the selected scRNA-seq cell\n",
    "    \"\"\"\n",
    "    \n",
    "    # Bug: the index in subsetted libsize different from the original count index!!!\n",
    "    cell_types = np.unique(df.index.get_level_values(1))\n",
    "    assert cell_type in cell_types, 'Unexpected cell type {}'.format(cell_type)\n",
    "    \n",
    "    # 'Mask out' indices other than the given cell type\n",
    "    mask = df.index.get_level_values(1) == cell_type\n",
    "    libsize = libsize_raw.copy()\n",
    "    libsize[~mask] = -np.inf   \n",
    "    \n",
    "    n = np.random.randint(n_nbr) if n_nbr > 0 else 0\n",
    "    idx = np.abs(libsize - l).argsort()[n]\n",
    "    expr = df.iloc[idx].to_numpy()\n",
    "    \n",
    "    return expr\n",
    "\n",
    "\n",
    "def fit_st_libsize(libsize, idxs):\n",
    "    \"\"\"\n",
    "    Fit library size from scRNA-seq data & sample expected library size \n",
    "    to each synthetic Spatial Transcriptomics spot\n",
    "    \"\"\"\n",
    "    \n",
    "    s, loc, scale = lognorm.fit(libsize)\n",
    "    l = pd.Series(\n",
    "        lognorm.rvs(s, loc=loc, scale=scale, size=len(idxs)),\n",
    "        index=idxs\n",
    "    )\n",
    "    \n",
    "    return np.round(l).astype(np.uint32)\n",
    "    \n",
    "\n",
    "def simulate_spots(df, count_df, expr_factors, l, ratio=0.16, verbose=False):\n",
    "    \"\"\"\n",
    "    Simulate spot x gene synthetic matrix given \n",
    "    (1). per-cell density; (2). library size for each spot\n",
    "    \"\"\"\n",
    "    n_spot = count_df.shape[0]\n",
    "    n_genes = df.shape[1]\n",
    "    libsize = df.sum(1).to_numpy().astype(np.float64)\n",
    "    \n",
    "    sims = np.zeros((n_spot, n_genes))\n",
    "    \n",
    "    for i, cell_types in enumerate(expr_factors):\n",
    "        if verbose and i % 100 == 0:\n",
    "            print('Simulating spot {}...'.format(i))\n",
    "            \n",
    "        l_s = l[i] # expected library size for given spot\n",
    "        spot_expr = np.zeros(n_genes)\n",
    "        \n",
    "        cell_count = count_df[cell_types].iloc[i]\n",
    "        n_cells = cell_count.sum().astype(np.uint32)\n",
    "        \n",
    "        # Sample expected library size to each single-cell index \n",
    "        # captured to current spot\n",
    "        # for now fix the same expected library size for all scRNA-seq to the given spot\n",
    "        \"\"\"\n",
    "        expected_libsize = poisson.rvs(\n",
    "            l[i], \n",
    "            size=n_cells\n",
    "        )\n",
    "        \"\"\"\n",
    "        \n",
    "        cell_type_freqs = np.hstack([\n",
    "            np.repeat(cell_type, count)\n",
    "            for (cell_type, count) in zip(cell_types, cell_count)\n",
    "        ])\n",
    "        \n",
    "        # for count, cell_type in zip(expected_libsize, cell_type_freqs):\n",
    "        for cell_type in cell_type_freqs:\n",
    "            ct_expr = find_sc_index(df, cell_type, libsize, l_s, n_nbr=0)\n",
    "            spot_expr += ct_expr\n",
    "            \n",
    "        spot_expr = np.round(spot_expr * ratio).astype(np.uint64)\n",
    "        sims[i] = spot_expr\n",
    "        \n",
    "    sims_df = pd.DataFrame(\n",
    "        sims,\n",
    "        index=count_df.index,\n",
    "        columns=df.columns\n",
    "    )\n",
    "    \n",
    "    return sims_df\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e65af681",
   "metadata": {},
   "source": [
    "- Sample library size $l_s$ for each synthetic spot from the r.v. fitted from scRNA-seq.<br>\n",
    "  ($l_s$ restricts the max. number of gene expression count for each spot $s$)<br>\n",
    "- Simulate synthetic expression ($X_{s,g}$):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30d69381",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get list of expressed (non-zero) cell type for each spot\n",
    "expressed_factors = find_nonzero_cts(cell_count_df)\n",
    "\n",
    "# Fit library size to each synthetic spot\n",
    "l_s = fit_st_libsize(\n",
    "    libsize=df_sc.values.sum(1), \n",
    "    idxs=cell_abundances_df.index\n",
    ")\n",
    "\n",
    "synth_df = simulate_spots(\n",
    "    df_sc,\n",
    "    cell_count_df,\n",
    "    expressed_factors,\n",
    "    l_s,\n",
    "    ratio=cap_ratio_per_spot,\n",
    "    verbose=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "557f94bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "synth_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c3e50cc",
   "metadata": {},
   "source": [
    "Display library size per spot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92f53e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(6, 5))\n",
    "sns.heatmap(synth_df.sum(1).to_numpy().reshape(50, -1))\n",
    "plt.axis('off')\n",
    "plt.title('Library size per spot')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b142da8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Comparison of real & synthetic ST library size histogram\n",
    "\n",
    "fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(10, 12))\n",
    "\n",
    "ax1.hist(adata.X.sum(1), bins=60, density=True, edgecolor='white')\n",
    "ax1.set_title('Library size - real ST matrix')\n",
    "\n",
    "ax2.hist(synth_df.sum(1), bins=60, density=True, edgecolor='white')\n",
    "ax2.set_title('Library size - synthetic ST matrix')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fcd3ad0",
   "metadata": {},
   "source": [
    "Save simulations to output\n",
    "- Spot x Gene synthetic matrix\n",
    "- Ground-truth cell count (density)\n",
    "- Ground-truth cell proportion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87c05555",
   "metadata": {},
   "outputs": [],
   "source": [
    "%ls ../../sc_transcriptome_ana/simulation/data_cell2loc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "003ad9a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save simulations to file\n",
    "out_path = '../../sc_transcriptome_ana/simulation/data_cell2loc/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73dab366",
   "metadata": {},
   "outputs": [],
   "source": [
    "synth_df.to_csv(\n",
    "    os.path.join(out_path, 'counts.st_synth.csv'),\n",
    "    index=True\n",
    ")\n",
    "\n",
    "cell_count_df.to_csv(\n",
    "    os.path.join(out_path, 'members.st_synth.csv'),\n",
    "    index=True\n",
    ")\n",
    "\n",
    "proportion_df.to_csv(\n",
    "    os.path.join(out_path, 'proportions.st_synth.csv'),\n",
    "    index=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b474309c",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b08003d5",
   "metadata": {},
   "source": [
    "#### [Benchmark] Check correlation between signature mean gexp. & ground-truth proportion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bab6fe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load signature gene dict\n",
    "\n",
    "# specify signature genes directory\n",
    "sig_genes_df = pd.read_csv('../../sc_transcriptome_ana/simulation/data/tnbc_signature.csv', header=[0])\n",
    "sig_genes_dict = sig_genes_df.to_dict(orient='list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58f84e69",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = '../../sc_transcriptome_ana/simulation/data_cell2loc/'\n",
    "\n",
    "synth_df = pd.read_csv(\n",
    "    os.path.join(out_path, 'counts.st_synth.csv'),\n",
    "    index_col=[0],\n",
    "    header=[0]\n",
    ")\n",
    "\n",
    "cell_count_df = pd.read_csv(\n",
    "    os.path.join(out_path, 'members.st_synth.csv'),\n",
    "    index_col=[0],\n",
    "    header=[0]\n",
    ")\n",
    "\n",
    "proportion_df = pd.read_csv(\n",
    "    os.path.join(out_path, 'proportions.st_synth.csv'),\n",
    "    index_col=[0],\n",
    "    header=[0]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a87f8bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parallel comparison with Stereoscope-like simulation (without normalization)\n",
    "\n",
    "stereo_sim_path = '../../sc_transcriptome_ana/simulation/data_stereo/'\n",
    "\n",
    "stereo_synth_df = pd.read_csv(\n",
    "    os.path.join(stereo_sim_path, 'counts.st_synth.csv'),\n",
    "    index_col=[0],\n",
    "    header=[0]\n",
    ")\n",
    "\n",
    "stereo_cell_count_df = pd.read_csv(\n",
    "    os.path.join(stereo_sim_path, 'members.st_synth.csv'),\n",
    "    index_col=[0],\n",
    "    header=[0]\n",
    ")\n",
    "\n",
    "stereo_proportion_df = pd.read_csv(\n",
    "    os.path.join(stereo_sim_path, 'proportions.st_synth.csv'),\n",
    "    index_col=[0],\n",
    "    header=[0]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e4e37f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def disp_density_gexp(expr_df, prop_df, sig_dict, normalize=False):\n",
    "    \"\"\"\n",
    "    Display relation of ground-truth cell-type proportion vs. \n",
    "    average of cell-type specific marker gene expression \n",
    "    \"\"\"\n",
    "    \n",
    "    cell_types = prop_df.columns\n",
    "    assert len(sig_dict) == len(cell_types) and \\\n",
    "           np.array_equal(sorted(sig_dict.keys()), sorted(cell_types)), \\\n",
    "        \"Unequal length of cell types in signature dictionary & ground-truth proportion dataframe\"\n",
    "    \n",
    "    expr_df_copy = expr_df.copy()\n",
    "    \n",
    "    if normalize:\n",
    "        med_libsize = expr_df_copy.sum(1).median()\n",
    "        expr_df_copy = expr_df_copy.div(expr_df_copy.sum(1), axis=0) * med_libsize\n",
    "    \n",
    "    # Average expression of cell-type specific marker genes\n",
    "    sig_expr = np.zeros(prop_df.shape) \n",
    "    for i, ct in enumerate(prop_df.columns):\n",
    "        sig_genes = np.intersect1d(sig_dict[ct], expr_df.columns)\n",
    "        expr = expr_df_copy[sig_genes].mean(1)\n",
    "        sig_expr[:, i] = expr\n",
    "        \n",
    "    sig_expr = sig_expr / np.expand_dims(sig_expr.sum(1), axis=1) # normalize expression to proportion\n",
    "    sig_expr_df = pd.DataFrame(\n",
    "        sig_expr,\n",
    "        index=expr_df.index,\n",
    "        columns = prop_df.columns\n",
    "    )\n",
    "    \n",
    "    # Plotting\n",
    "    r = 0\n",
    "    c = 0    \n",
    "    nrow = np.ceil(len(cell_types) // 3).astype(np.uint8)\n",
    "    fig, axes = plt.subplots(nrow, 3, figsize=(15, 15))\n",
    "    \n",
    "    for i, ct in enumerate(cell_types):\n",
    "        if c % 3 == 0 and c > 0:\n",
    "            r += 1\n",
    "            c = 0\n",
    "        \n",
    "        xx = prop_df[ct].to_numpy()\n",
    "        yy = sig_expr_df[ct].to_numpy()\n",
    "        \n",
    "        ax = axes[r, c]\n",
    "        ax.scatter(xx, yy, alpha=0.5, marker='.', cmap=plt.cm.Greens)\n",
    "        ax.set_xlim([0, 1])\n",
    "        ax.set_ylim([0, 1])\n",
    "        ax.set_xlabel('Ground-truth proportion')\n",
    "        ax.set_ylabel('Mean sig gexp')\n",
    "        ax.set_title(ct)\n",
    "        \n",
    "        c += 1\n",
    "        \n",
    "    plt.tight_layout()\n",
    "    # plt.suptitle('Cell2Location simulation (w/ spatial dependencies)', y=1.05, fontsize=20)\n",
    "    plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d24573c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cell2location simulation (w/ spatial dependencies)\n",
    "disp_density_gexp(synth_df, proportion_df, sig_genes_dict, normalize=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da026955",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Stereoscope simulation (without spatial dependencies)\n",
    "disp_density_gexp(stereo_synth_df, stereo_proportion_df, sig_genes_dict, normalize=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee7812ca",
   "metadata": {},
   "source": [
    "---"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
